### PR TITLE
feat(tracing): spans not sent to agent after tracer shut down

### PR DESCRIPTION
**Pull request by mabdinur**
Warns users when ``Tracer.trace()`` or ``Tracer.start_span()`` is called after a tracer is shutdown.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Comment by mabdinur** [comment]
This change logs a warning when ``tracer.trace()`` is called while the agent writer is stopped. All tests that stop the agent writer, generating traces and then asserts specific logs will fail :(

File: tests/contrib/gevent/test_monkeypatch.py [hunk 1]
```diff
@@ -20,17 +20,22 @@ def test_gevent_warning(monkeypatch):
     assert b"RuntimeWarning: Loading ddtrace before using gevent monkey patching" in subp.stderr.read()
 
 
-@pytest.mark.subprocess
-def test_gevent_auto_patching():
```
Comment (mabdinur) [inline tests/contrib/gevent/test_monkeypatch.py:24]: This refactor prevents this test from logging ``Tracer was shutdown. Spans generated after a tracer is shutdown will not be sent.``

File: tests/contrib/gevent/test_monkeypatch.py [hunk 2]
```diff
@@ -20,17 +20,22 @@ def test_gevent_warning(monkeypatch):
...
 
-    ddtrace.patch_all()
-    # Patch on import
-    import gevent  # noqa
+    out, err, status, pid = run_python_code_in_subprocess(
+        """
+import ddtrace
+ddtrace.patch_all()
+# Patch on import
+import gevent  # noqa
```
Comment (brettlangdon) [inline tests/contrib/gevent/test_monkeypatch.py:30]: you don't need `  # noqa` in a code block like this

File: tests/contrib/gevent/test_monkeypatch.py (hunk 1 above)
Comment (brettlangdon) [reply tests/contrib/gevent/test_monkeypatch.py:24]: I am confused by this comment, the changes you made have it where this test does *not* log this warning, and therefore it fails?

Comment (mabdinur) [reply tests/contrib/gevent/test_monkeypatch.py:24]: The ``pyest.mark.subprocess`` annotation calles [run_function_from_file](https://github.com/DataDog/dd-trace-py/blob/1.x/tests/conftest.py#L160) and then fails the assertion on [line 197](https://github.com/DataDog/dd-trace-py/blob/1.x/tests/conftest.py#L160) after checking the output of ``ddtrace-run test.py``. I couldn't reproduce this error locally but here's the error in [ci](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/15672/workflows/e8cf1079-1d80-4423-be99-8e38f547be84/jobs/1080110). Using ``run_python_code_in_subprocess or ddtrace_run_python_code_in_subprocess`` fixtures instead prevented this warning log (Now the AgentWriter is not shutdown before tracer.trace()) is called. I don't think patching or accessing the ``GeventContextProvider`` should produce traces so I have no idea why this test is failing.

**Review by majorgreys — COMMENTED** [review]
There's a potential race condition here where a span is started after the writer is stopped but before the start method is changed.

I am inclined to have this check handled by `AgentWriter.write`. 

Also, should this be more than just a warning? If the writer service has been stopped, spans should not be queued up for writing. However, we do not distinguish between a service stopped and a service that has not yet been started.

File: ddtrace/internal/writer.py [hunk 3]
```diff
@@ -488,13 +489,17 @@ def write(self, spans=None):
...
             # Start the AgentWriter on first write.
             try:
                 if self.status != service.ServiceStatus.RUNNING:
                     self.start()
             except service.ServiceStatusError:
                 pass
+            self._start_writer = True
+
+        if self.status == service.ServiceStatus.STOPPED:
+            log.warning("Tracer was shutdown. Spans generated after a tracer is shutdown will not be sent.")
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:502]: should we return here or let the writer encode spans?

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -460,7 +461,21 @@ def _child_after_fork(self):
...
+        self,
+        name,  # type: str
+        child_of=None,  # type: Optional[Union[Span, Context]]
+        service=None,  # type: Optional[str]
+        resource=None,  # type: Optional[str]
+        span_type=None,  # type: Optional[str]
+        activate=False,  # type: bool
+    ):
+        # type: (...) -> Span
+        if isinstance(self._writer, AgentWriter) and self._writer.status == ServiceStatus.STOPPED:
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:474]:
```suggestion
```

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: tests/tracer/test_tracer.py [hunk 5]
```diff
@@ -560,6 +560,22 @@ def test_tracer_shutdown_timeout():
     t._writer.stop.assert_called_once_with(timeout=2)
 
 
+def test_send_traces_after_shutdown():
+    t = ddtrace.Tracer()
+    assert isinstance(t._writer, AgentWriter)
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_tracer.py:565]:
```suggestion
```

File: releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+---
+features:
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml:2]: Maybe just use `upgrade` and remind again that this behaviour isn't supported/encouraged?

Feature is a bit weird to me for this change.

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: ddtrace/tracer.py [hunk 7]
```diff
@@ -460,7 +460,20 @@ def _child_after_fork(self):
...
+        self,
+        name,  # type: str
+        child_of=None,  # type: Optional[Union[Span, Context]]
+        service=None,  # type: Optional[str]
+        resource=None,  # type: Optional[str]
+        span_type=None,  # type: Optional[str]
+        activate=False,  # type: bool
+    ):
+        # type: (...) -> Span
+        log.warning("Tracer was shutdown. Spans generated after a tracer is shutdown will not be sent to the agent.")
```
Comment (majorgreys) [inline ddtrace/tracer.py:473]:
```suggestion
        log.warning("Spans started after tracer has been shut down will not sent to the Datadog Agent.")
```

File: releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml (hunk 6 above)
Comment (majorgreys) [reply releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml:2]: Yes, this should be an upgrade as it's a breaking change with 1.0.

File: tests/contrib/gevent/test_monkeypatch.py [hunk 8]
```diff
@@ -20,17 +20,22 @@ def test_gevent_warning(monkeypatch):
     assert b"RuntimeWarning: Loading ddtrace before using gevent monkey patching" in subp.stderr.read()
 
 
-@pytest.mark.subprocess
-def test_gevent_auto_patching():
-    import ddtrace
+def test_gevent_auto_patching(ddtrace_run_python_code_in_subprocess):
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply tests/contrib/gevent/test_monkeypatch.py:23]: Reverting this change will cause this test to fail: https://app.circleci.com/jobs/github/DataDog/dd-trace-py/1090603?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link. Here's a discussion about this issue https://github.com/DataDog/dd-trace-py/pull/3483#discussion_r835525411

File: ddtrace/tracer.py [hunk 9]
```diff
@@ -460,7 +460,20 @@ def _child_after_fork(self):
...
+        self,
+        name,  # type: str
+        child_of=None,  # type: Optional[Union[Span, Context]]
+        service=None,  # type: Optional[str]
+        resource=None,  # type: Optional[str]
+        span_type=None,  # type: Optional[str]
+        activate=False,  # type: bool
+    ):
+        # type: (...) -> Span
+        log.warning("Spans started after tracer has been shut down will not sent to the Datadog Agent.")
```
Comment (majorgreys) [inline ddtrace/tracer.py:473]:
```suggestion
        log.warning("Spans started after the tracer has been shut down will not be sent to the Datadog Agent.")
```

File: releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+---
+upgrade:
+  - |
+    Log warning when :py:meth:`ddtrace.tracer.start_span` is called after a tracer is shutdown.
```
Comment (majorgreys) [inline releasenotes/notes/add-warning-after-shutdown-tracer-dd8799467c4d29c8.yaml:4]:
```suggestion
    Spans started after the tracer has been shut down will no longer be sent to the Datadog Agent.
```

File: tests/contrib/gevent/test_monkeypatch.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mabdinur) [reply tests/contrib/gevent/test_monkeypatch.py]: I reverted this change in the latest commit. I'll keep an eye on ci

**Comment by majorgreys** [comment]
@Mergifyio backport 1.0

File: tests/tracer/test_tracer.py [hunk 11]
```diff
@@ -560,6 +560,21 @@ def test_tracer_shutdown_timeout():
...
+    t = ddtrace.Tracer()
+    t.shutdown()
+
+    with mock.patch.object(logging.Logger, "warning") as mock_logger:
+        with t.trace("something"):
+            pass
+
+    mock_logger.assert_has_calls(
+        [
+            mock.call("Spans started after tracer has been shut down will not sent to the Datadog Agent."),
```
Comment (mabdinur) [inline tests/tracer/test_tracer.py:573]:
```suggestion
            mock.call("Spans started after the tracer has been shut down will not be sent to the Datadog Agent."),
```

Warning changed in a previous commit

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

File: tests/contrib/gevent/test_monkeypatch.py [hunk 12]
```diff
@@ -24,7 +24,8 @@ def test_gevent_warning(monkeypatch):
 def test_gevent_auto_patching():
     import ddtrace
 
-    ddtrace.patch_all()
+    # Disable tracing sqlite3 as it is used by coverage
+    ddtrace.patch_all(sqlite3=False)
```
Comment (Kyle-Verhoog) [inline tests/contrib/gevent/test_monkeypatch.py:28]: 👍 this has come up before and probably worth solving in general for CI

https://github.com/DataDog/dd-trace-py/blob/c437ad25a916f0abc8d2b44e7470cf659e7ade3e/tests/integration/test_integration.py#L774-L777

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**